Fix first and last when records are not loaded in a sorted relationship

GermanDZ opened the pull request:
### Summary

Fix first, last, and the other ordinal finders (second, second_to_last, …) on a sorted relation whose order has ties (multiple records sharing the same value for the ordering column).

A test was added and is failing with current implementation, for all databases tests, i.e.:

<img width="1129" alt="image" src="https://user-images.githubusercontent.com/570991/69701553-7a196680-10ed-11ea-84d4-543ffa36c4d5.png">

When the relation isn't already loaded, these finders push the work to the database using the explicit order plus a LIMIT/OFFSET. If the explicit order isn't unique, the
  database is free to break ties arbitrarily — so first and last could return the same record, and loaded vs. non-loaded relations could disagree.

```ruby
  # owner.pets is ordered by `name`, and both pets are named "kiwi"
  owner.pets.first  # ORDER BY pets.name DESC LIMIT 1
  owner.pets.last   # ORDER BY pets.name ASC  LIMIT 1   → could be the same row
```

#### Solution
   
In the shared ordered_relation (used by every ordinal finder), when an explicit order is present we append the model's default order columns (primary_key, implicit_order_column, or query_constraints) as a secondary sort to disambiguate ties:

```ruby
  owner.pets.first  # ORDER BY pets.name DESC, pets.pet_id ASC  LIMIT 1
  owner.pets.last   # ORDER BY pets.name ASC,  pets.pet_id DESC LIMIT 1
```

The secondary columns are only added when they aren't already part of the order (including orders given as raw SQL strings), so no redundant sort columns are emitted.

#### Related issues

Relationship to #39525:
This is orthogonal to the recently merged default_order (#39525): default_order chooses an order when none is specified, whereas this change adds a deterministic tie-breaker when an explicit order is specified but may have ties.
  
Probably solving this issue #36475?
No, after reproducing the case with the fix applies is still failing.

p8 commented:
Hi @GermanDZ, Thanks for creating this PR. There also is a newer related ticket: #39525

jonathanhefner commented:
I agree with @p8 that #39525 would be a better approach to solve this, though it wouldn't be an "automatic" solution.

One problem with this PR is that it doesn't address the order of `first`, nor any of the other ordinal finders (e.g. `owner.pets.second`).

A larger problem, though, is that it affects `last` unconditionally.  For example, `Part.order(:serial_number).last` will now always use `ORDER BY "parts"."serial_number" DESC, "parts"."id" DESC`, which is both unexpected and a potential performance issue.

GermanDZ commented:
@p8 / @jonathanhefner I've just added my failing test case (just a cherry-pick of my commit) to the #39525 PR and is still failing.

We should consider mix both solutions in a single PR? Maybe we could wait until the other PR is merged and then adapt this PR to fix "my" problem on top of the previous (better) solution.

WDYT?

GermanDZ commented:
Waiting for merging of #39525 to rebase and rework this solution

GermanDZ commented:
Still waiting for #39525 to rebase and rework this solution

GermanDZ commented:
Still waiting for #39525 to rebase and rework this solution

GermanDZ commented:
Still waiting for #39525 to rebase and rework this solution

GermanDZ commented:
Still waiting for https://github.com/rails/rails/pull/39525 to rebase and rework this solution

GermanDZ committed:
Add failing test showing the bad behaviour of last and first in ordered relationships

GermanDZ committed:
Fix `first` and `last` behavior to be deterministic

rafaelfranca commented:
https://github.com/rails/rails/pull/39525 was merged finally. Can you rebase/rework this branch?

GermanDZ commented:
@rafaelfranca Thanks!

Rebased onto current main (which now includes #39525) and reworked. The branch is now two clean commits: a failing test, then the fix.

@jonathanhefner Revisiting your two points from before, since the rework addresses both:

  1. first and the other ordinal finders. The fix now lives in the shared ordered_relation, which every ordinal finder routes through (`first`/`second` via `find_nth`, `last`/`second_to_last` via `find_nth_from_last`). So the deterministic tie-breaker is applied symmetrically across all of them, not just last:

```ruby
  Pet.where(owner_id: 3).order("pets.name desc")
  .first           # ORDER BY pets.name desc, "pets"."pet_id" ASC
  .second          # ORDER BY pets.name desc, "pets"."pet_id" ASC  ... OFFSET
  .last            # ORDER BY pets.name ASC,  "pets"."pet_id" DESC
  .second_to_last  # ORDER BY pets.name ASC,  "pets"."pet_id" DESC ... OFFSET
```

  I've added test coverage for second/second_to_last alongside the original first/last cases.

  2. The unconditional secondary sort on `last`. The original objection was that last was being modified asymmetrically. Now first and last get the same treatment, so the inconsistency between them (and between loaded vs. non-loaded relations) is what's being fixed by appending the model's order columns as a secondary sort is the mechanism that makes the result deterministic when the explicit order has ties. The extra column is the primary key (or implicit_order_column/query_constraints), which is unique, so it only breaks ties and is cheap on an indexed key.

  To avoid redundancy, the tie-breaker columns are only appended when they aren't already part of the order, including order values given as raw SQL strings:

```ruby
  Pet.order("pets.name desc, pets.pet_id desc").first
  # ORDER BY pets.name desc, pets.pet_id desc   ← no redundant pet_id appended
```

**Note**: this is orthogonal to #39525's default_order, which applies only when no order is specified; this handles the case where an explicit order may have ties.

GermanDZ committed:
Merge branch 'main' into last-and-first-in-active-record-sorted-collection-must-work-as-expected

GermanDZ committed:
Merge branch 'main' into last-and-first-in-active-record-sorted-collection-must-work-as-expected
